File: src/mission/ServerScriptService/Zombies/Gunslinger.lua
local ReplicatedStorage = game:GetService("ReplicatedStorage")
local ServerScriptService = game:GetService("ServerScriptService")

local TakeDamage = require(ServerScriptService.Shared.TakeDamage)

local GunslingerZombieEffect = ReplicatedStorage.Remotes.Zombies.GunslingerZombieEffect

local Gunslinger = {}
Gunslinger.__index = Gunslinger

Gunslinger.AttackCooldown = 2
Gunslinger.AttackRange = 25
Gunslinger.Model = "Gunslinger"

function Gunslinger.new()
	return setmetatable({}, Gunslinger)
end

function Gunslinger:AfterSpawn()
	local instance = self.instance
	local gun = instance.Gun

	local idleAnimation = self:LoadAnimation(gun.Animations.Idle)
	idleAnimation.Priority = Enum.AnimationPriority.Idle
	idleAnimation.Looped = true
	idleAnimation:Play()

	local aimAnimation = self:LoadAnimation(gun.Animations.Aim)
	aimAnimation.Priority = Enum.AnimationPriority.Idle
	aimAnimation.Looped = true

	self.aimAnimation = aimAnimation
	self.reloadAnimation = self:LoadAnimation(gun.Animations.Reload)
	self.shootAnimation = self:LoadAnimation(gun.Animations.AimShoot)

	self.aliveMaid:GiveTask(GunslingerZombieEffect.OnServerEvent:connect(function(player)
		TakeDamage(player, self:GetDamageAgainst(player))
	end))
end

function Gunslinger:Attack()
	self.aimAnimation:Play()
	GunslingerZombieEffect:FireAllClients(self.instance)
	wait(self:GetScale("ActivationTime"))
	if not self.alive then return end
	GunslingerZombieEffect:FireAllClients(self.instance, true)
	self.aimAnimation:Stop()
	self.shootAnimation:Play()

	return true
end

function Gunslinger:GetAttackCooldown()
	return Gunslinger.AttackCooldown + self:GetScale("ActivationTime")
end

return Gunslinger

File: src/mission/ServerScriptService/Zombies/Gunslinger.ts
import { ReplicatedStorage } from "@rbxts/services"
import Common from "./Common"
import TakeDamage from "shared/ServerScriptService/TakeDamage"

const GunslingerZombieEffect = ReplicatedStorage.Remotes.Zombies.GunslingerZombieEffect

class Gunslinger extends Common {
	static Model = "Gunslinger"

	AfterSpawn() {
		this.AttackRange = this.GetScale("Range") / 2

		this.aliveMaid.GiveTask(GunslingerZombieEffect.OnServerEvent.Connect((player, zombie) => {
			if (zombie === this.instance) {
				TakeDamage(player, this.GetScale("Damage"))
			}
		}))

		const range = new Instance("NumberValue")
		range.Name = "Range"
		range.Value = this.GetScale("Range")
		range.Parent = this.instance.WaitForChild("Gun")
	}

	Attack() {
		GunslingerZombieEffect.FireAllClients(this.instance, this.aggroFocus)
		this.instance.Humanoid.WalkSpeed = 0
		wait(this.GetScale("ActivationTime"))
		if (this.alive) {
			this.instance.Humanoid.WalkSpeed = this.GetSpeed()
		}

		return true
	}

	GetAttackCooldown() {
		return this.GetScale("Cooldown") + this.GetScale("ActivationTime")
	}
}

export = Gunslinger
